Applicative expansion of string_with_vars

rgrinberg opened the pull request:
Allow the function used to expand in String_with_vars to have an
applicative effect.

aalekseyev commented on src/dune/string_with_vars.ml:
```diff
@@ -229,68 +229,108 @@ end
+end
+
+module Make (A : Applicative_intf.S1) = struct
+  module App = Applicative.Make (A)
```
For many applicatives `all` can be more efficient than fold of `let+`.
I think it's fine to have `all` as a field of `A` directly instead of deriving it here. (it's not a big deal at this point, though)

rgrinberg replied:
Agreed. However, I'd prefer working on applying this change to expanding actions within the applicative. This is more important for the overall progress I feel.

aalekseyev commented:
Looks good to me!

snowleopard commented on src/stdune/applicative_intf.ml:
```diff
@@ -0,0 +1,9 @@
+  val ( let+ ) : 'a t -> ('a -> 'b) -> 'b t
+
+  val ( and+ ) : 'a t -> 'b t -> ('a * 'b) t
+end
```
I suggest some changes to this interface:
* Rename `pure` to `return` (since our existing applicatives follow this convention).
* Add `map` and `map2` -- they are quite pervasive and will certainly be needed at some point.
* Move `let+` and `and+` into a sub-module `O` (again, our usual convention), and also to `Make` so that they can be obtained automatically.

rgrinberg replied:
Isn't it a bit awkward to require all these helper functions? I.e. If I want to instantiate some code with a concrete applicative, I'd rather provide the smallest interface possible to it.

I don't have any strong preferences here, so feel free to push changes and tweak the interface as you wish

rgrinberg replied:
Wouldn't it be a bit annoying to provide all these auxiliary functions when instantiating something like `String_with_vars.Make`? Or perhaps, I'm misunderstanding something. Feel free to push the changes you have in mind to this branch by the way. I don't feel too strongly about this stuff.

aalekseyev replied:
I think eventually it's good to have (at least) two versions of applicative interface: a convenient one with at least `return`, `map`, `both`, `module O`, `all` etc., and a minimal one to pass to `Applicative.Make`. I think most applicative-generic code should take the convenient one, not the minimal one, so that the custom efficient versions  of things like `all` can be used.

In the end I don't think we'll need more than a few applicatives so the difficulty of implementing the "big" applicative interface doesn't frighten me.

snowleopard commented on src/dune/string_with_vars.mli:
```diff
@@ -114,14 +114,24 @@ val fold_vars : t -> init:'a -> f:(Var.t -> 'a -> 'a) -> 'a
-  -> dir:Path.t
-  -> f:Value.t list option expander
-  -> 'a Partial.t
+module type S = sig
```
Can you add some comments here? What do these two functions do, and why do we need to parameterise over the applicative?

rgrinberg replied:
Sure, I'll add some comments. The basic idea is that we'd like to accumulate a `Build.t` while expanding.

snowleopard replied:
Thanks! The (almost complete) lack of comments in Dune makes it really hard for me to understand various key mechanisms.

rgrinberg commented:
This PR now addresses all the reviewers comments

snowleopard commented:
> This PR now addresses all the reviewers comments

Not all: I still see no comments which you promised -- please add them, or we will forever be scratching our heads looking at the code we wrote two years ago and no longer understand.

rgrinberg commented:
I’ve added one comment explaining why we need the applicative. I have another PR where I’m adding some comments to both this module and Expander. Check out #3169 and feel free to ask questions on which parts are unclear.
On Feb 20, 2020, 9:45 AM +0000, Andrey Mokhov <notifications@github.com>, wrote:
> > This PR now addresses all the reviewers comments
> Not all: I still see no comments which you promised -- please add them, or we will forever be scratching our heads looking at the code we wrote two years ago and no longer understand.
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub, or unsubscribe.

snowleopard commented:
What I really miss when working with Dune is a (brief) top-level summary for each module, explaining its overall purpose and pointing to 2-3 key functions. Since you are refactoring this module right now and have the most up-to-date knowledge of its internals, you are the best person to write one! I don't really mind which PR it comes from, but hope it won't get lost.
